File: sol/113.cpp
struct Solution {
    void dfs(TreeNode *node, int curr, vector<int> &rec, vector<vector<int>> &res){
        curr -= node->val;
        rec.push_back(node->val);
        if (node->left == nullptr && node->right == nullptr && curr == 0){
            res.push_back(rec);
        }
        if (node->left != nullptr){
            this->dfs(node->left, curr, rec, res);
        }
        if (node->right != nullptr){
            this->dfs(node->right, curr, rec, res);
        }
        rec.pop_back();
    }
    inline vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int> rec;
        vector<vector<int>> res;
        if (root == nullptr){
            return res;
        }
        this->dfs(root, targetSum, rec, res);
        return res;
    }
};

File: sol/113/113.py
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        if root is None:
            return []
        res = []
        rec = []
        def dfs(node, curr):
            curr -= node.val
            rec.append(node.val)
            if node.left is None and node.right is None and curr == 0:
                res.append(rec[:])
            if node.left is not None:
                dfs(node.left, curr)
            if node.right is not None:
                dfs(node.right, curr)
            rec.pop()
        dfs(root, targetSum)
        return res
